#!/usr/bin/env cwl-runner

cwlVersion: v1.0

class: CommandLineTool

requirements:
  - class: InlineJavascriptRequirement

hints:
  - class: DockerRequirement
    dockerPull: ACCOUNT/gtex_v10:VERSION

baseCommand: [/src/run_rnaseqc.py]

inputs:
  - id: gene_annotation_gtf
    type: File
    inputBinding:
      position: 1
    doc: Collapsed gene annotation model in GTF format

  - id: input_file_bam
    type: File
    inputBinding:
      position: 2
    secondaryFiles:
      - .bai
    doc: Alignment BAM file generated by STAR with the corresponding index file. |
         Need to be sorted by coordinates

  - id: output_prefix
    type: string
    default: "out"
    inputBinding:
      position: 3
    doc: Prefix for output files

  - id: strandedness
    type: string
    default: null
    inputBinding:
      position: 4
      prefix: --stranded
    doc: Strandedness for stranded libraries. |
         "rf" or "fr"

outputs:
  - id: output_gene_tpm_tsv
    type: File
    outputBinding:
      glob: $(inputs.output_prefix + ".gene_tpm.tsv")

  - id: output_tar_gz
    type: File
    outputBinding:
      glob: $(inputs.output_prefix + ".all.tar.gz")

doc: |
  Run RNA-SeQC to generate quality metrics and gene TPM quantification
