#!/usr/bin/env cwl-runner

cwlVersion: v1.0

class: CommandLineTool

requirements:
  - class: InlineJavascriptRequirement

hints:
  - class: DockerRequirement
    dockerPull: ACCOUNT/gtex_v10:VERSION

baseCommand: [run_RSEM.sh]

inputs:
  - id: input_transcriptome_bam
    type: File
    inputBinding:
      position: 1
    doc: Transcriptome BAM file generated by STAR

  - id: genome_reference_rsem
    type: File
    inputBinding:
      position: 2
    doc: RSEM reference index files generated with rsem-prepare-reference. |
         Expect a compressed archive. |
         The file prefix inside the archive need to be rsem_reference

  - id: is_stranded
    type: string
    default: 'false'
    inputBinding:
      position: 3
    doc: Set to 'true' if the protocol is strand-specific

outputs:
  - id: output_isoforms_tsv
    type: File
    outputBinding:
      glob: isoforms.tsv

  - id: output_genes_tsv
    type: File
    outputBinding:
      glob: genes.tsv

doc: |
  Run RSEM for transcripts quantification
